### ISSUE TITLE
gRPC Interface not returning `rows_before_limit_at_least` using format JSON

### ISSUE THREAD

**Issue by lesandie** [ClickHouse/ClickHouse#38806]
gRPC Interface is not returning the same field structure as native or http interface. Tested v22.3.

```sql
$clickhouse-client --host localhost --port 9000 -mn
ClickHouse client version 22.3.7.28 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 22.3.7 revision 54455.

laptop :) select * from system.one limit 1 format JSON;

SELECT *
FROM system.one
LIMIT 1
FORMAT JSON

Query id: bc846be7-948a-4668-95ef-d47c8d5e24dc

{
	"meta":
	[
		{
			"name": "dummy",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"dummy": 0
		}                                                                                                                                   
	],

	"rows": 1,

	"rows_before_limit_at_least": 1,

	"statistics":
	{
		"elapsed": 0.000291421,
		"rows_read": 1,
		"bytes_read": 1
	}
}

1 rows in set. Elapsed: 0.001 sec.
```

```sql
./clickhouse-grpc-client.py
Connecting to localhost:9100 as user default.
Connected to ClickHouse server version 22.3.7.28 via gRPC protocol in 0.004273414611816406.
grpc :) select * from system.one limit 1 format JSON;
{
	"meta":
	[
		{
			"name": "dummy",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"dummy": 0
		}
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.000573447,
		"rows_read": 0,
		"bytes_read": 0
	}
}

Elapsed: 0.0037772655487060547 sec.
```
A probable cause is that there is no `rows_before_limit_at_least` defined property in proto file message `Result ` structure:

https://github.com/ClickHouse/ClickHouse/blob/59236d60c9d46bfa130d670e21e2eff7d1560e89/src/Server/grpc_protos/clickhouse_grpc.proto#L192

**Comment by arthurpassos** [comment]
`clickhouse-grpc-client` prints out `Result::output` and ignores `Result::stats`, see [code](https://github.com/ClickHouse/ClickHouse/blob/70066986e97f2684dc81b859db42002265e67be0/utils/grpc-client/clickhouse-grpc-client.py#L161-L169). `Stats` does contain `rows_before_limit`, see [Stats structure](https://github.com/ClickHouse/ClickHouse/blob/59236d60c9d46bfa130d670e21e2eff7d1560e89/src/Server/grpc_protos/clickhouse_grpc.proto#L181).

That means the data is being returned by the server, but it's not being printed by the client. You are reporting this issue because you want the data to be printed by `clickhouse-grpc-client` or because you want to consume it via a different client? If the latter, one option would be to grab it from `stats`.

**Comment by filimonov** [comment]
But if the GRPC requests data in a certain format, and server does that formatting - then why server does not add that field to the output? 

For example when server prints that output using HTTP - it fills the subkey:

```
echo 'select * from system.one limit 1 format JSON' | curl http://localhost:8123/ --data-binary @-
```

**Comment by filimonov** [comment]
So i guess the question is more about why clickhouse server does not enter that if
https://github.com/ClickHouse/ClickHouse/blob/22e11aef2ded232859651602636f846eed0f97a6/src/Formats/JSONUtils.cpp#L585

**Comment by lesandie** [comment]
Also gRPC interface is returning the elapsed time but not rows_read and bytes_read:

```json
Connecting to localhost:9100 as user default.
Connected to ClickHouse server version 22.8.5.29 via gRPC protocol in 0.007185697555541992.
grpc :) select * from system.one limit 1 format JSON
{
	"meta":
	[
		{
			"name": "dummy",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"dummy": 0
		}
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.000383378,
		"rows_read": 0,
		"bytes_read": 0
	}
}

Elapsed: 0.003061532974243164 sec.

grpc :) 
```

HTTP interface:

```json
echo 'select * from system.one limit 1 format JSON' | curl http://localhost:8123/ --data-binary @-

{
	"meta":
	[
		{
			"name": "dummy",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"dummy": 0
		}
	],

	"rows": 1,

	"rows_before_limit_at_least": 1,

	"statistics":
	{
		"elapsed": 0.000278001,
		"rows_read": 1,
		"bytes_read": 1
	}
}
```
NATIVE

```json
SELECT *
FROM system.one
LIMIT 1
FORMAT JSON

Query id: ade23ce6-cf98-44eb-9b94-8bdc639c2a51

{
	"meta":
	[
		{
			"name": "dummy",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"dummy": 0
		}                                                                                                                   
	],

	"rows": 1,

	"rows_before_limit_at_least": 1,

	"statistics":
	{
		"elapsed": 0.000282607,
		"rows_read": 1,
		"bytes_read": 1
	}
}

1 row in set. Elapsed: 0.001 sec. 
```

Also the protobuf file in the Results message references the Stats:
https://github.com/ClickHouse/ClickHouse/blob/59236d60c9d46bfa130d670e21e2eff7d1560e89/src/Server/grpc_protos/clickhouse_grpc.proto#L210

But `rows_before_limits` is rendered outside the stats message in other interfaces. So it seems the problem is more complicated: also gRPC interface is not rendering results for all of the stats properties but elapsed (rows_read, bytes_read are 0)

Also I've commented the above line in the `clickhouse_grpc.proto` and recompiled the protobuf file. Results are the same. So either the server is returning the data correctly and the client is not rendering it or the server is not branching by that if condition. I'm more inclined to the first option. I'll try to test with another grpc cli to check for consistency in results.

**Comment by lesandie** [comment]
@arthurpassos What do you think?. Maybe the way to go is changing the clickhouse grpc client as there are no other options/clients to test this, apart from build one.

**Comment by arthurpassos** [comment]
> So i guess the question is more about why clickhouse server does not enter that if
> 
> https://github.com/ClickHouse/ClickHouse/blob/22e11aef2ded232859651602636f846eed0f97a6/src/Formats/JSONUtils.cpp#L585

IIRC, It does not enter that branch because `IOutputFormat::setRowsBeforeLimit` is not called for the JSON output format, it is only called for a `LazyOutputFormat` instance. `clickhouse-client` renders that field because it calls it, but that's client side. See `clickhouse-client` [IOutputFormat::setRowsBeforeLimit call](https://github.com/ClickHouse/ClickHouse/blob/22e11aef2ded232859651602636f846eed0f97a6/src/Client/ClientBase.cpp#L471).

**Comment by arthurpassos** [comment]
> But if the GRPC requests data in a certain format, and server does that formatting - then why server does not add that field to the output?
> 
> For example when server prints that output using HTTP - it fills the subkey:
> 
> ```
> echo 'select * from system.one limit 1 format JSON' | curl http://localhost:8123/ --data-binary @-
> ```

What do you mean by fills subkey?

**Comment by arthurpassos** [comment]
> @arthurpassos What do you think?. Maybe the way to go is changing the clickhouse grpc client as there are no other options/clients to test this, apart from build one.

I am not sure, I guess it depends on the customer necessity/ motivation. Bottom line is that the field is being returned in the response, but it's not being rendered. If the customer necessity is to see that field through grpc-client, we have two paths to follow. Either change the position where it's returned in the server or render it on client side like `clickhouse-client` does. If the customer necessity is to consume this via a different grpc client (i.e his own application with grpc), then there are two paths forward: client renders it from the `stats` structure or we change server to send it outside `stats` structure.

Does that make sense? I have just come back from PTO, so I might be a bit off and trying to remember things lol

**Comment by lesandie** [comment]
Yes, it does make sense 😄 . I'd go with the python client change for the moment, because the client only wanted a CLI with the same features as other interfaces like HTTP or Native.
